client/asset/dcr/coin_selection_test.go

```diff
@@ -8,110 +8,6 @@ import (
 	walletjson "decred.org/dcrwallet/v2/rpc/jsonrpc/types"
 )
 
-func Test_subsetLargeBias(t *testing.T) {
-	amt := uint64(10e8)
-	newU := func(amt float64) *compositeUTXO {
-		return &compositeUTXO{
-			rpc: &walletjson.ListUnspentResult{Amount: amt},
-		}
-	}
-	tests := []struct {
-		name  string
-		utxos []*compositeUTXO
-		want  []*compositeUTXO
-	}{
-		{
-			"1,3 exact",
-			[]*compositeUTXO{newU(1), newU(8), newU(9)},
-			[]*compositeUTXO{newU(1), newU(9)},
-		},
-		{
-			"subset large bias",
-			[]*compositeUTXO{newU(1), newU(3), newU(6), newU(7)},
-			[]*compositeUTXO{newU(3), newU(7)},
-		},
-		{
-			"subset large bias",
-			[]*compositeUTXO{newU(1), newU(3), newU(5), newU(7), newU(8)},
-			[]*compositeUTXO{newU(3), newU(8)},
-		},
-		{
-			"insufficient",
-			[]*compositeUTXO{newU(1), newU(8)},
-			nil,
-		},
-		{
-			"all exact",
-			[]*compositeUTXO{newU(1), newU(9)},
-			[]*compositeUTXO{newU(1), newU(9)},
-		},
-	}
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			if got := subsetLargeBias(amt, tt.utxos); !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("subset() = %v, want %v", got, tt.want)
-			}
-		})
-	}
-}
-
-func Test_subsetSmallBias(t *testing.T) {
-	amt := uint64(10e8)
-	newU := func(amt float64) *compositeUTXO {
-		return &compositeUTXO{
-			rpc: &walletjson.ListUnspentResult{Amount: amt},
-		}
-	}
-	tests := []struct {
-		name  string
-		utxos []*compositeUTXO
-		want  []*compositeUTXO
-	}{
-		{
-			"1,3",
-			[]*compositeUTXO{newU(1), newU(8), newU(9)},
-			[]*compositeUTXO{newU(8), newU(9)},
-		},
-		{
-			"subset",
-			[]*compositeUTXO{newU(1), newU(9), newU(11)},
-			[]*compositeUTXO{newU(1), newU(9)},
-		},
-		{
-			"subset small bias",
-			[]*compositeUTXO{newU(1), newU(3), newU(6), newU(7)},
-			[]*compositeUTXO{newU(1), newU(3), newU(6)},
-		},
-		{
-			"subset small bias",
-			[]*compositeUTXO{newU(1), newU(3), newU(5), newU(7), newU(8)},
-			[]*compositeUTXO{newU(5), newU(7)},
-		},
-		{
-			"ok nil",
-			[]*compositeUTXO{newU(1), newU(8)},
-			nil,
-		},
-		{
-			"two, over",
-			[]*compositeUTXO{newU(5), newU(7), newU(11)},
-			[]*compositeUTXO{newU(5), newU(7)},
-		},
-		{
-			"insufficient",
-			[]*compositeUTXO{newU(1), newU(8)},
-			nil,
-		},
-	}
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			if got := subsetSmallBias(amt, tt.utxos); !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("subset() = %v, want %v", got, tt.want)
-			}
-		})
-	}
-}
-
 func Test_leastOverFund(t *testing.T) {
 	amt := uint64(10e8)
 	newU := func(amt float64) *compositeUTXO {
```

Comment:
Thanks for this!

I guess the test cases for the dumber leastOverFund were too easy?  We chatted a bit about replacing the algo and it sounded like the new one got better results ~1/2 the time.  Would it require much larger sets to test a case where it's a better result?

Reply:
The better results ~1/2 the time was with the dynamic programming solution which only worked with very small amounts of UTXOs. When I tested this one with large amounts of UTXOs and compared it with the old solution, it got better results every single time.

The two solutions (this vs small and large bias subset) are equivalent, but the old one tries 2 random possibilities and this one tries 1000.

Reply:
Tests well.  ~16ms for a set of 4000 UTXOs between 0 and 2 DCR in search of a subset totaling 10DCR

```go
func Test_subsetFund(t *testing.T) {
	amt := uint64(10e8)
	s := make([]*compositeUTXO, 4000)
	newU := func(amt float64) *compositeUTXO {
		return &compositeUTXO{
			rpc: &walletjson.ListUnspentResult{Amount: amt},
		}
	}
	for i := range s {
		v := rand.Float64() + float64(rand.Int63n(2))
		v = math.Round(v*1e8) / 1e8
		fmt.Println(v)
		s[i] = newU(v)
	}
	sort.Slice(s, func(i, j int) bool {
		return s[i].rpc.Amount < s[i].rpc.Amount
	})
	r := subsetWithLeastSumGreaterThan(amt, s)
	fmt.Println(len(r), sumUTXOs(r))
}
```